pallets/investments/src/lib.rs

```diff
@@ -0,0 +1,898 @@
+			.payout_asset_redeem
+			.checked_add(
+				&fulfillment
+					.price
+					.checked_mul_int(fulfillment.redeem.mul_floor(order.redeem))
+					.ok_or(ArithmeticError::Overflow)?,
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		Ok(())
+	}
+
+	pub fn acc_remaining_redeem(
+		collection: &mut Collection<T::Amount>,
+		fulfillment: &FulfillmentWithPrice<T::BalanceRatio>,
+		order: &Order<T::Amount, T::OrderId>,
+	) -> DispatchResult {
+		collection.remaining_asset_redeem = collection
+			.remaining_asset_redeem
+			.checked_sub(&fulfillment.redeem.mul_floor(order.redeem))
+			.ok_or(ArithmeticError::Underflow)?;
+
+		Ok(())
+	}
+
+	pub fn acc_remaining_invest(
+		collection: &mut Collection<T::Amount>,
+		fulfillment: &FulfillmentWithPrice<T::BalanceRatio>,
+		order: &Order<T::Amount, T::OrderId>,
+	) -> DispatchResult {
+		collection.remaining_asset_invest = collection
+			.remaining_asset_invest
+			.checked_sub(&fulfillment.redeem.mul_floor(order.invest))
+			.ok_or(ArithmeticError::Underflow)?;
+
+		Ok(())
+	}
+}
+
+impl<T: Config> InvestmentManager for Pallet<T>
+where
+	<T::Accountant as AssetAccountant<T::AccountId>>::AssetInfo:
+		AssetProperties<T::AccountId, Currency = CurrencyOf<T>>,
+{
+	type Error = DispatchError;
+	type AssetId = T::InvestmentId;
+	type Orders = TotalOrder<T::Amount>;
+	type OrderId = T::OrderId;
+	type Fulfillment = FulfillmentWithPrice<T::BalanceRatio>;
+
+	fn orders(asset_id: Self::AssetId) -> Result<(Self::OrderId, Self::Orders), Self::Error> {
+		let order = ActiveOrder::<T>::get(&asset_id);
+		let order_id = OrderId::<T>::get(&asset_id);
+
+		InProcessingOrders::<T>::insert(&asset_id, &order_id, order.clone());
+		OrderId::<T>::insert(
+			&asset_id,
+			&order_id
+				.checked_add(&One::one())
+				.ok_or(ArithmeticError::Overflow)?,
+		);
+
+		Self::deposit_event(Event::OrderInProcessing {
+			asset_id,
+			order_id,
+			total_order: order.clone(),
+		});
+
+		Ok((order_id, order))
+	}
+
+	fn fulfillment(
+		order_id: Self::OrderId,
+		asset_id: Self::AssetId,
+		fulfillment: Self::Fulfillment,
+	) -> Result<(), DispatchError> {
+		InProcessingOrders::<T>::try_mutate(
+			&asset_id,
+			&order_id,
+			|maybe_orders| -> DispatchResult {
+				let orders = maybe_orders
+					.as_ref()
+					.ok_or(Error::<T>::OrderNotInProcessing)?;
+
+				let invest = fulfillment.invest.mul_floor(orders.invest);
+				// The orders for redemptions are denominated on a per
+				// asset basis. Hence, we need to convert it the amount
+				// of payment_currency that is redeemed by multiplying it
+				// with the price per asset unit.
+				let redeem = fulfillment.redeem.mul_floor(
+					fulfillment
+						.price
+						.checked_mul_int(orders.redeem)
+						.ok_or(ArithmeticError::Overflow)?,
+				);
+				let asset_account = AssetAccount {
+					asset_id: asset_id.clone(),
+				}
+				.into_account_truncating();
+				let info = T::Accountant::info(asset_id.clone())?;
+
+				if invest >= redeem {
```

Comment:
Should this not be `invest > redeem`? If `invest == redeem`, then there's nothing to transfer.